// @midival/core/dist/mpe/output/MPEMidivalOutput.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _MPEMidivalOutput_lowerZone, _MPEMidivalOutput_upperZone;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MPEMidivalOutput = void 0;
const MIDIValOutput_1 = require("../../MIDIValOutput");
const MPEOutputZone_1 = require("./MPEOutputZone");
class MPEMidivalOutput {
    get lowerZone() {
        return __classPrivateFieldGet(this, _MPEMidivalOutput_lowerZone, "f");
    }
    get upperZone() {
        return __classPrivateFieldGet(this, _MPEMidivalOutput_upperZone, "f");
    }
    constructor(output, options) {
        this.options = options;
        _MPEMidivalOutput_lowerZone.set(this, null);
        _MPEMidivalOutput_upperZone.set(this, null);
        this.midivalOutput = new MIDIValOutput_1.MIDIValOutput(output);
        this.midivalOutput.initializeMPE(this.options.lowerZoneSize || 0, this.options.upperZoneSize || 0, 10);
        if (this.options.lowerZoneSize) {
            __classPrivateFieldSet(this, _MPEMidivalOutput_lowerZone, new MPEOutputZone_1.MPEOutputZone(1, [2, 2 + this.options.lowerZoneSize], this.midivalOutput), "f");
        }
        if (this.options.upperZoneSize) {
            __classPrivateFieldSet(this, _MPEMidivalOutput_upperZone, new MPEOutputZone_1.MPEOutputZone(16, [15 - this.options.upperZoneSize, 15], this.midivalOutput), "f");
        }
    }
    disconnect() {
        this.midivalOutput.initializeMPE(0, 0);
        __classPrivateFieldSet(this, _MPEMidivalOutput_lowerZone, null, "f");
        __classPrivateFieldSet(this, _MPEMidivalOutput_upperZone, null, "f");
    }
}
exports.MPEMidivalOutput = MPEMidivalOutput;
_MPEMidivalOutput_lowerZone = new WeakMap(), _MPEMidivalOutput_upperZone = new WeakMap();
