// @midival/core/dist/wrappers/inputs/BrowserMIDIInput.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserMIDIInput = void 0;
class BrowserMIDIInput {
    constructor(input) {
        this.input = input;
    }
    onMessage(fn) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.input.open();
            this.input.addEventListener("midimessage", fn);
            return () => {
                this.input.removeEventListener("midimessage", fn);
            };
        });
    }
    get id() {
        return this.input.id;
    }
    get name() {
        return this.input.name;
    }
    get manufacturer() {
        return this.input.manufacturer;
    }
}
exports.BrowserMIDIInput = BrowserMIDIInput;
